File: packaging/python-identity/tests/test_dev_identity.py
"""Local-dev identity: with no ShinyHub proxy in front, current_user can
return a synthetic Identity from SHINYHUB_IDENTITY_DEV_* env vars so every app
does not have to invent its own mock. Guards make it impossible to trigger
under a real deployment: it only activates when no token arrived AND no
verification key exists (ShinyHub always injects SHINYHUB_IDENTITY_KEY)."""

import time

import jwt  # PyJWT
import pytest

from shinyhub_identity import Identity, current_user

KEY = bytes.fromhex("00112233445566778899aabbccddeeff00112233445566778899aabbccddeeff")
SLUG = "sales-dashboard"


@pytest.fixture(autouse=True)
def clean_env(monkeypatch):
    for var in (
        "SHINYHUB_IDENTITY_KEY",
        "SHINYHUB_APP_SLUG",
        "SHINYHUB_IDENTITY_DEV_USER",
        "SHINYHUB_IDENTITY_DEV_GROUPS",
        "SHINYHUB_IDENTITY_DEV_EMAIL",
        "SHINYHUB_IDENTITY_DEV_NAME",
        "SHINYHUB_IDENTITY_DEV_ROLE",
    ):
        monkeypatch.delenv(var, raising=False)
    return monkeypatch


def mint_for(slug, key):
    now = int(time.time())
    return jwt.encode(
        {"iss": "shinyhub", "sub": "42", "aud": slug, "preferred_username": "alice",
         "role": "admin", "groups": [], "iat": now, "exp": now + 300},
        key,
        algorithm="HS256",
    )


def test_dev_user_env_yields_synthetic_identity(clean_env):
    clean_env.setenv("SHINYHUB_IDENTITY_DEV_USER", "devlin")
    clean_env.setenv("SHINYHUB_IDENTITY_DEV_GROUPS", "team-a, team-b")
    clean_env.setenv("SHINYHUB_IDENTITY_DEV_EMAIL", "devlin@example.com")
    clean_env.setenv("SHINYHUB_IDENTITY_DEV_NAME", "Devlin Example")
    user = current_user({})
    assert isinstance(user, Identity)
    assert user.username == "devlin"
    assert user.user_id == "devlin"
    assert user.groups == ("team-a", "team-b")
    assert user.email == "devlin@example.com"
    assert user.name == "Devlin Example"
    assert user.groups_truncated is False


def test_dev_identity_is_marked_in_claims(clean_env):
    clean_env.setenv("SHINYHUB_IDENTITY_DEV_USER", "devlin")
    user = current_user({})
    assert user.claims["dev"] is True


def test_dev_role_defaults_to_viewer_and_is_overridable(clean_env):
    clean_env.setenv("SHINYHUB_IDENTITY_DEV_USER", "devlin")
    assert current_user({}).role == "viewer"
    clean_env.setenv("SHINYHUB_IDENTITY_DEV_ROLE", "admin")
    assert current_user({}).role == "admin"


def test_no_dev_env_means_anonymous(clean_env):
    assert current_user({}) is None


def test_dev_identity_never_activates_under_a_real_deployment(clean_env):
    # ShinyHub always injects SHINYHUB_IDENTITY_KEY into worker envs, so its
    # presence means "deployed": the dev identity must not mask a real
    # verification failure or a genuinely anonymous visitor there.
    clean_env.setenv("SHINYHUB_IDENTITY_DEV_USER", "devlin")
    clean_env.setenv("SHINYHUB_IDENTITY_KEY", KEY.hex())
    assert current_user({}) is None


def test_dev_identity_does_not_shadow_a_real_token(clean_env):
    clean_env.setenv("SHINYHUB_IDENTITY_DEV_USER", "devlin")
    headers = {"x-shinyhub-identity-token": mint_for(SLUG, KEY)}
    user = current_user(headers, key=KEY, slug=SLUG)
    assert user.username == "alice"


def test_dev_identity_skipped_when_key_passed_explicitly(clean_env):
    # An explicit key= means the caller is exercising the real verify path
    # (e.g. in tests); a missing token there must stay anonymous.
    clean_env.setenv("SHINYHUB_IDENTITY_DEV_USER", "devlin")
    assert current_user({}, key=KEY, slug=SLUG) is None


def test_dev_groups_parsing_ignores_blanks(clean_env):
    clean_env.setenv("SHINYHUB_IDENTITY_DEV_USER", "devlin")
    clean_env.setenv("SHINYHUB_IDENTITY_DEV_GROUPS", " team-a ,, team-b , ")
    assert current_user({}).groups == ("team-a", "team-b")

File: packaging/r-identity/tests/testthat/test-dev-identity.R
# Local-dev identity: with no ShinyHub proxy in front, current_user can return
# a synthetic identity from SHINYHUB_IDENTITY_DEV_* env vars. Guards make it
# impossible to trigger under a real deployment: it only activates when no
# token arrived AND no verification key exists (ShinyHub always injects
# SHINYHUB_IDENTITY_KEY into app processes).

test_that("dev env vars yield a synthetic identity", {
  with_env(list(
    SHINYHUB_IDENTITY_KEY = NA,
    SHINYHUB_IDENTITY_DEV_USER = "devlin",
    SHINYHUB_IDENTITY_DEV_GROUPS = "team-a, team-b",
    SHINYHUB_IDENTITY_DEV_EMAIL = "devlin@example.com",
    SHINYHUB_IDENTITY_DEV_NAME = "Devlin Example",
    SHINYHUB_IDENTITY_DEV_ROLE = NA
  ), {
    u <- current_user(list(request = list()))
    expect_s3_class(u, "shinyhub_identity")
    expect_equal(u$username, "devlin")
    expect_equal(u$user_id, "devlin")
    expect_equal(u$role, "viewer")
    expect_equal(u$groups, c("team-a", "team-b"))
    expect_equal(u$email, "devlin@example.com")
    expect_equal(u$name, "Devlin Example")
    # The marker lives in the raw claims, where it cannot be mistaken for a
    # verified field of the identity itself.
    expect_true(isTRUE(u$claims$dev))
  })
})

test_that("dev role is overridable", {
  with_env(list(
    SHINYHUB_IDENTITY_KEY = NA,
    SHINYHUB_IDENTITY_DEV_USER = "devlin",
    SHINYHUB_IDENTITY_DEV_ROLE = "admin"
  ), {
    expect_equal(current_user(list(request = list()))$role, "admin")
  })
})

test_that("dev identity never activates when the real key is present", {
  with_env(list(
    SHINYHUB_IDENTITY_KEY = strrep("00", 32),
    SHINYHUB_IDENTITY_DEV_USER = "devlin"
  ), {
    expect_null(current_user(list(request = list())))
  })
})

test_that("dev identity does not shadow a real token", {
  with_env(list(
    SHINYHUB_IDENTITY_KEY = NA,
    SHINYHUB_IDENTITY_DEV_USER = "devlin"
  ), {
    session <- list(request = list(HTTP_X_SHINYHUB_IDENTITY_TOKEN = dmint()))
    u <- current_user(session, key = dkey, slug = dslug)
    expect_equal(u$username, "alice")
  })
})

test_that("dev identity is skipped when a key is passed explicitly", {
  with_env(list(
    SHINYHUB_IDENTITY_KEY = NA,
    SHINYHUB_IDENTITY_DEV_USER = "devlin"
  ), {
    expect_null(current_user(list(request = list()), key = dkey, slug = dslug))
  })
})

test_that("absent dev env means anonymous", {
  with_env(list(
    SHINYHUB_IDENTITY_KEY = NA,
    SHINYHUB_APP_SLUG = NA,
    SHINYHUB_IDENTITY_DEV_USER = NA
  ), {
    expect_null(current_user(list(request = list())))
  })
})
